gluten-core/src/main/scala/io/glutenproject/utils/AdaptiveSparkPlanUtil.scala

```diff
@@ -17,28 +17,55 @@
 
 package io.glutenproject.utils
 
-import org.apache.spark.sql.execution.SparkPlan
-import org.apache.spark.sql.execution.exchange.Exchange
+import io.glutenproject.execution.CoalesceBatchesExec
+import org.apache.spark.sql.execution.{ColumnarShuffleExchangeAdaptor, ColumnarShuffleExchangeExec, SparkPlan}
+import org.apache.spark.sql.execution.exchange.{Exchange, ShuffleExchangeExec}
 import org.apache.spark.sql.internal.SQLConf
 
 /**
  * Mostly ported from spark source code for checking whether a plan supports adaptive.
+ * See InsertAdaptiveSparkPlan#supportAdaptive in vanilla spark.
  * Since spark-3.2, AQE can work for DPP, so no need to exclude DPP plan in the check.
  * This part of code may need update for supporting higher versions of spark.
  */
 object AdaptiveSparkPlanUtil {
 
   def sanityCheck(plan: SparkPlan): Boolean = plan.logicalLink.isDefined
 
+  def supportAdaptive(plan: SparkPlan): Boolean = {
+    SQLConf.get.adaptiveExecutionEnabled &&
+        (sanityCheck(plan) &&
+            !plan.logicalLink.exists(_.isStreaming) &&
+            plan.children.forall(supportAdaptive))
+  }
+
   def supportAdaptiveWithExchangeConsidered(plan: SparkPlan): Boolean = {
     // Only QueryStage will have Exchange as Leaf Plan
     val isLeafPlanExchange = plan match {
       case _: Exchange => true
       case _ => false
     }
-    isLeafPlanExchange || (SQLConf.get.adaptiveExecutionEnabled &&
-        (sanityCheck(plan) &&
-            !plan.logicalLink.exists(_.isStreaming) &&
-            plan.children.forall(supportAdaptiveWithExchangeConsidered)))
+    isLeafPlanExchange || supportAdaptive(plan)
+  }
+
+  /**
+   * Generate a columnar plan for shuffle exchange.
+   *
+   * @param plan             the spark plan of shuffle exchange.
+   * @param child            the child of shuffle exchange.
+   * @param removeHashColumn whether the hash column should be removed.
+   * @return a columnar shuffle exchange.
+   */
+  def genColumnarShuffleExchange(plan: ShuffleExchangeExec,
```

Comment:
It looks weird to put the `genColumnarShuffleExchange` in AdaptiveSparkPlanUtil, because there's no direct correlation between exchange and AQE. But it's ok with me if it's only temporary solution.

Reply:
Yes, it is better to move it outside this file. Will do this in another PR. Thanks!